// @tma.js/transformers/dist/index.js
import { throwifyFpFn as g } from "@tma.js/toolkit";
import { pipe as c, string as n, check as S, transform as p, union as A, instance as D, parse as L, looseObject as l, optional as i, number as w, boolean as m, integer as N, date as O, record as I, is as U, unknown as z } from "valibot";
function E(t, r, e, u, a, s, o, R, J) {
  switch (arguments.length) {
    case 1:
      return t;
    case 2:
      return r(t);
    case 3:
      return e(r(t));
    case 4:
      return u(e(r(t)));
    case 5:
      return a(u(e(r(t))));
    case 6:
      return s(a(u(e(r(t)))));
    case 7:
      return o(s(a(u(e(r(t))))));
    case 8:
      return R(o(s(a(u(e(r(t)))))));
    case 9:
      return J(R(o(s(a(u(e(r(t))))))));
    default: {
      for (var y = arguments[0], _ = 1; _ < arguments.length; _++)
        y = arguments[_](y);
      return y;
    }
  }
}
var V = function(t) {
  return t._tag === "Left";
}, q = function(t) {
  return { _tag: "Left", left: t };
}, v = function(t) {
  return { _tag: "Right", right: t };
}, Q = q, f = v, x = function(t) {
  return function(r) {
    return j(r) ? r : f(t(r.right));
  };
}, j = V, G = function(t, r) {
  try {
    return f(t());
  } catch (e) {
    return Q(r(e));
  }
};
function M(t) {
  const r = (e) => {
    const u = {};
    return new URLSearchParams(e).forEach((a, s) => {
      const o = u[s];
      Array.isArray(o) ? o.push(a) : o === void 0 ? u[s] = a : u[s] = [o, a];
    }), L(t, u);
  };
  return c(
    A([n(), D(URLSearchParams)]),
    S((e) => {
      try {
        return r(e), !0;
      } catch {
        return !1;
      }
    }, "The value doesn't match required schema"),
    p(r)
  );
}
function H(t) {
  return c(
    n(),
    S((r) => {
      try {
        return JSON.parse(r), !0;
      } catch {
        return !1;
      }
    }, "Input is not a valid JSON value"),
    p(JSON.parse),
    t
  );
}
function h(t) {
  return c(n(), H(t));
}
function W(t) {
  return c(
    A([n(), D(URLSearchParams)]),
    M(t)
  );
}
function d(t) {
  return (r) => t.test(r);
}
const K = d(/^#[\da-f]{3}$/i), X = d(/^#[\da-f]{4}$/i), Y = d(/^#[\da-f]{6}$/i), Z = d(/^#[\da-f]{8}$/i);
function k(t) {
  return [Y, Z, K, X].some((r) => r(t));
}
function $(t) {
  let r = "#";
  for (let e = 0; e < t.length - 1; e += 1)
    r += t[1 + e].repeat(2);
  return r;
}
function b(t) {
  const r = t.replace(/\s/g, "").toLowerCase();
  if (/^#[\da-f]{3}$/i.test(r))
    return f($(r.toLowerCase() + "f"));
  if (/^#[\da-f]{4}$/i.test(r))
    return f($(r.toLowerCase()));
  if (/^#[\da-f]{6}$/i.test(r))
    return f(r.toLowerCase() + "ff");
  if (/^#[\da-f]{8}$/i.test(r))
    return f(r.toLowerCase());
  const e = r.match(/^rgb\((\d{1,3}),(\d{1,3}),(\d{1,3})\)$/) || r.match(/^rgba\((\d{1,3}),(\d{1,3}),(\d{1,3}),(\d{1,3})\)$/);
  return e ? f(
    e.slice(1).reduce(
      (u, a) => u + parseInt(a, 10).toString(16).padStart(2, "0"),
      "#"
    ).padEnd(9, "f")
  ) : Q(new Error(`Value "${t}" does not satisfy any of known RGB formats.`));
}
function tt(t) {
  return E(
    b(t),
    x((r) => r.slice(0, 7))
  );
}
const ct = g(tt), ft = g(b);
function rt() {
  return l({
    id: w(),
    photo_url: i(n()),
    type: n(),
    title: n(),
    username: i(n())
  });
}
function P() {
  return l({
    added_to_attachment_menu: i(m()),
    allows_write_to_pm: i(m()),
    first_name: n(),
    id: w(),
    is_bot: i(m()),
    is_premium: i(m()),
    last_name: i(n()),
    language_code: i(n()),
    photo_url: i(n()),
    username: i(n())
  });
}
function et() {
  return l({
    auth_date: c(
      n(),
      p((t) => new Date(Number(t) * 1e3)),
      O()
    ),
    can_send_after: i(c(n(), p(Number), N())),
    chat: i(h(rt())),
    chat_type: i(n()),
    chat_instance: i(n()),
    hash: n(),
    query_id: i(n()),
    receiver: i(h(P())),
    start_param: i(n()),
    signature: n(),
    user: i(h(P()))
  });
}
function F() {
  return W(et());
}
function B() {
  return I(
    n(),
    c(
      A([n(), w()]),
      p((t) => typeof t == "number" ? `#${(t & 16777215).toString(16).padStart(6, "0")}` : t),
      S(k)
    )
  );
}
function nt() {
  const t = i(c(n(), p((r) => r === "1")));
  return l({
    tgWebAppBotInline: t,
    tgWebAppData: i(F()),
    tgWebAppDefaultColors: i(h(B())),
    tgWebAppFullscreen: t,
    tgWebAppPlatform: n(),
    tgWebAppShowSettings: t,
    tgWebAppStartParam: i(n()),
    tgWebAppThemeParams: h(B()),
    tgWebAppVersion: n()
  });
}
function T() {
  return W(nt());
}
function pt(t) {
  try {
    return U(T(), t);
  } catch {
    return !1;
  }
}
function ht() {
  return l({
    eventType: n(),
    eventData: i(z())
  });
}
function at(t) {
  return G(
    () => L(F(), t),
    (r) => r
  );
}
function it(t) {
  return G(
    () => L(T(), t),
    (r) => r
  );
}
const lt = g(at), mt = g(it);
function C(t, r) {
  return r || (r = (e, u) => JSON.stringify(u)), new URLSearchParams(
    Object.entries(t).reduce((e, [u, a]) => (Array.isArray(a) ? e.push(...a.map((s) => [u, String(s)])) : a != null && e.push([
      u,
      a instanceof Date ? (a.getTime() / 1e3 | 0).toString() : typeof a == "string" || typeof a == "number" ? String(a) : typeof a == "boolean" ? a ? "1" : "0" : r(u, a)
    ]), e), [])
  ).toString();
}
function ut(t) {
  return C(t);
}
function gt(t) {
  return C(t, (r, e) => r === "tgWebAppData" ? ut(e) : JSON.stringify(e));
}
export {
  et as initData,
  rt as initDataChat,
  F as initDataQuery,
  P as initDataUser,
  k as isAnyRGB,
  pt as isLaunchParamsQuery,
  Y as isRGB,
  Z as isRGBA,
  X as isRGBAShort,
  K as isRGBShort,
  nt as launchParams,
  T as launchParamsQuery,
  ht as miniAppsMessage,
  lt as parseInitDataQuery,
  at as parseInitDataQueryFp,
  mt as parseLaunchParamsQuery,
  it as parseLaunchParamsQueryFp,
  h as pipeJsonToSchema,
  W as pipeQueryToSchema,
  ut as serializeInitDataQuery,
  gt as serializeLaunchParamsQuery,
  B as themeParams,
  ct as toRGB,
  tt as toRGBFp,
  ft as toRGBFull,
  b as toRGBFullFp
};
