// @tma.js/toolkit/dist/index.js
import { BetterPromise as y } from "better-promises";
function ar(r) {
  return r.replace(/[A-Z]/g, (n) => `-${n.toLowerCase()}`);
}
function E(r) {
  return r.replace(/[A-Z]/g, (n) => `_${n.toLowerCase()}`);
}
function ir(r) {
  return Object.entries(r).reduce((n, [t, e]) => (n[E(t)] = e, n), {});
}
function F(r) {
  return r.replace(/_[a-z]/g, (n) => n[1].toUpperCase());
}
function I(r) {
  return Object.entries(r).reduce((n, [t, e]) => (n[F(t)] = e, n), {});
}
function v(r) {
  const n = I(r);
  for (const t in n) {
    const e = n[t];
    e && typeof e == "object" && !(e instanceof Date) && (n[t] = Array.isArray(e) ? e.map(v) : v(e));
  }
  return n;
}
function cr(r) {
  return r.replace(/_([a-z])/g, (n, t) => `-${t.toLowerCase()}`);
}
function m(r) {
  return `tapps/${r}`;
}
function sr(r, n) {
  sessionStorage.setItem(m(r), JSON.stringify(n));
}
function fr(r) {
  const n = sessionStorage.getItem(m(r));
  try {
    return n ? JSON.parse(n) : void 0;
  } catch {
  }
}
function lr(...r) {
  const n = r.flat(1);
  return [
    n.push.bind(n),
    () => {
      n.forEach((t) => {
        t();
      });
    }
  ];
}
// @__NO_SIDE_EFFECTS__
function hr(r, n) {
  n || (n = {});
  const {
    textColor: t,
    bgColor: e,
    shouldLog: a
  } = n, u = a === void 0 ? !0 : a, i = typeof u == "boolean" ? () => u : u, s = (f, o, ...c) => {
    if (o || i()) {
      const l = "font-weight:bold;padding:0 5px;border-radius:100px", [x, O, j] = {
        log: ["#0089c3", "white", "INFO"],
        error: ["#ff0000F0", "white", "ERR"],
        warn: ["#D38E15", "white", "WARN"]
      }[f];
      console[f](
        `%c${j} ${Intl.DateTimeFormat("en-GB", {
          hour: "2-digit",
          minute: "2-digit",
          second: "2-digit",
          fractionalSecondDigits: 3,
          timeZone: "UTC"
        }).format(/* @__PURE__ */ new Date())}%c %c${r}`,
        `${l};background-color:${x};color:${O}`,
        "",
        `${l};${t ? `color:${t};` : ""}${e ? `background-color:${e}` : ""}`,
        ...c
      );
    }
  };
  return [
    ["log", "forceLog"],
    ["warn", "forceWarn"],
    ["error", "forceError"]
  ].reduce((f, [o, c]) => (f[o] = s.bind(void 0, o, !1), f[c] = s.bind(void 0, o, !0), f), {});
}
var P = function(r, n, t) {
  if (t || arguments.length === 2) for (var e = 0, a = n.length, u; e < a; e++)
    (u || !(e in n)) && (u || (u = Array.prototype.slice.call(n, 0, e)), u[e] = n[e]);
  return r.concat(u || Array.prototype.slice.call(n));
};
function d(r, n, t, e, a, u, i, s, f) {
  switch (arguments.length) {
    case 1:
      return r;
    case 2:
      return function() {
        return n(r.apply(this, arguments));
      };
    case 3:
      return function() {
        return t(n(r.apply(this, arguments)));
      };
    case 4:
      return function() {
        return e(t(n(r.apply(this, arguments))));
      };
    case 5:
      return function() {
        return a(e(t(n(r.apply(this, arguments)))));
      };
    case 6:
      return function() {
        return u(a(e(t(n(r.apply(this, arguments))))));
      };
    case 7:
      return function() {
        return i(u(a(e(t(n(r.apply(this, arguments)))))));
      };
    case 8:
      return function() {
        return s(i(u(a(e(t(n(r.apply(this, arguments))))))));
      };
    case 9:
      return function() {
        return f(s(i(u(a(e(t(n(r.apply(this, arguments)))))))));
      };
  }
}
function h(r, n, t, e, a, u, i, s, f) {
  switch (arguments.length) {
    case 1:
      return r;
    case 2:
      return n(r);
    case 3:
      return t(n(r));
    case 4:
      return e(t(n(r)));
    case 5:
      return a(e(t(n(r))));
    case 6:
      return u(a(e(t(n(r)))));
    case 7:
      return i(u(a(e(t(n(r))))));
    case 8:
      return s(i(u(a(e(t(n(r)))))));
    case 9:
      return f(s(i(u(a(e(t(n(r))))))));
    default: {
      for (var o = arguments[0], c = 1; c < arguments.length; c++)
        o = arguments[c](o);
      return o;
    }
  }
}
var w = function(r, n) {
  var t = typeof r == "number" ? function(e) {
    return e.length >= r;
  } : r;
  return function() {
    var e = Array.from(arguments);
    return t(arguments) ? n.apply(this, e) : function(a) {
      return n.apply(void 0, P([a], e, !1));
    };
  };
}, R = function(r) {
  return r._tag === "Left";
}, $ = function(r) {
  return { _tag: "Left", left: r };
}, _ = function(r) {
  return { _tag: "Right", right: r };
}, C = $, k = _, A = R, D = function(r, n) {
  return function(t) {
    return A(t) ? r(t.left) : n(t.right);
  };
}, g = D;
function K(r) {
  return d(k, r.of);
}
function U(r) {
  return d(C, r.of);
}
function W(r) {
  return function(n, t) {
    return r.chain(n, function(e) {
      return A(e) ? r.of(e) : t(e.right);
    });
  };
}
function B(r) {
  return function(n, t) {
    return function(e) {
      return r.map(e, g(n, t));
    };
  };
}
var L = function(r, n) {
  return h(r, M(n));
}, N = function(r, n) {
  return h(r, V(n));
}, M = function(r) {
  return function(n) {
    return function() {
      return Promise.resolve().then(n).then(r);
    };
  };
}, V = function(r) {
  return function(n) {
    return function() {
      return Promise.all([Promise.resolve().then(n), Promise.resolve().then(r)]).then(function(t) {
        var e = t[0], a = t[1];
        return e(a);
      });
    };
  };
}, p = function(r) {
  return function() {
    return Promise.resolve(r);
  };
}, Z = /* @__PURE__ */ w(2, function(r, n) {
  return function() {
    return Promise.resolve().then(r).then(function(t) {
      return n(t)();
    });
  };
}), S = "Task", z = {
  URI: S,
  map: L
}, T = {
  of: p
}, G = {
  URI: S,
  map: L,
  of: p,
  ap: N,
  chain: Z
}, J = function(r, n, t, e) {
  function a(u) {
    return u instanceof t ? u : new t(function(i) {
      i(u);
    });
  }
  return new (t || (t = Promise))(function(u, i) {
    function s(c) {
      try {
        o(e.next(c));
      } catch (l) {
        i(l);
      }
    }
    function f(c) {
      try {
        o(e.throw(c));
      } catch (l) {
        i(l);
      }
    }
    function o(c) {
      c.done ? u(c.value) : a(c.value).then(s, f);
    }
    o((e = e.apply(r, n || [])).next());
  });
}, q = function(r, n) {
  var t = { label: 0, sent: function() {
    if (u[0] & 1) throw u[1];
    return u[1];
  }, trys: [], ops: [] }, e, a, u, i;
  return i = { next: s(0), throw: s(1), return: s(2) }, typeof Symbol == "function" && (i[Symbol.iterator] = function() {
    return this;
  }), i;
  function s(o) {
    return function(c) {
      return f([o, c]);
    };
  }
  function f(o) {
    if (e) throw new TypeError("Generator is already executing.");
    for (; i && (i = 0, o[0] && (t = 0)), t; ) try {
      if (e = 1, a && (u = o[0] & 2 ? a.return : o[0] ? a.throw || ((u = a.return) && u.call(a), 0) : a.next) && !(u = u.call(a, o[1])).done) return u;
      switch (a = 0, u && (o = [o[0] & 2, u.value]), o[0]) {
        case 0:
        case 1:
          u = o;
          break;
        case 4:
          return t.label++, { value: o[1], done: !1 };
        case 5:
          t.label++, a = o[1], o = [0];
          continue;
        case 7:
          o = t.ops.pop(), t.trys.pop();
          continue;
        default:
          if (u = t.trys, !(u = u.length > 0 && u[u.length - 1]) && (o[0] === 6 || o[0] === 2)) {
            t = 0;
            continue;
          }
          if (o[0] === 3 && (!u || o[1] > u[0] && o[1] < u[3])) {
            t.label = o[1];
            break;
          }
          if (o[0] === 6 && t.label < u[1]) {
            t.label = u[1], u = o;
            break;
          }
          if (u && t.label < u[2]) {
            t.label = u[2], t.ops.push(o);
            break;
          }
          u[2] && t.ops.pop(), t.trys.pop();
          continue;
      }
      o = n.call(r, t);
    } catch (c) {
      o = [6, c], a = 0;
    } finally {
      e = u = 0;
    }
    if (o[0] & 5) throw o[1];
    return { value: o[0] ? o[1] : void 0, done: !0 };
  }
}, H = /* @__PURE__ */ U(T), Q = /* @__PURE__ */ K(T), X = p, b = /* @__PURE__ */ B(z), Y = b, rr = function(r, n) {
  return function() {
    return J(void 0, void 0, void 0, function() {
      var t;
      return q(this, function(e) {
        switch (e.label) {
          case 0:
            return e.trys.push([0, 2, , 3]), [4, r().then(_)];
          case 1:
            return [2, e.sent()];
          case 2:
            return t = e.sent(), [2, $(n(t))];
          case 3:
            return [
              2
              /*return*/
            ];
        }
      });
    });
  };
}, nr = /* @__PURE__ */ w(2, W(G)), tr = nr;
function er(r) {
  const n = (t) => {
    throw t;
  };
  return typeof r == "function" ? y.resolve(h(r, b(n, (t) => t))()) : h(r, g(n, (t) => t));
}
function gr(r) {
  return Object.assign(
    (...n) => er(r(...n)),
    r
  );
}
const ur = Object.assign(
  (r, n) => h(
    rr(
      () => new y((t, e, a) => r(
        (u) => t(k(u)),
        (u) => t(C(u)),
        a
      ), n),
      (t) => t
    ),
    tr(g(H, Q))
  ),
  {
    fn: (r, n) => ur((t, e, a) => {
      const u = r(a);
      h(
        typeof u == "function" ? u : X(u),
        Y(e, t)
      )();
    }, n)
  }
);
export {
  ur as BetterTaskEither,
  ar as camelToKebab,
  E as camelToSnake,
  ir as camelToSnakeObjKeys,
  lr as createCbCollector,
  hr as createLogger,
  v as deepSnakeToCamelObjKeys,
  fr as getStorageValue,
  sr as setStorageValue,
  F as snakeToCamel,
  I as snakeToCamelObjKeys,
  cr as snakeToKebab,
  er as throwifyAnyEither,
  gr as throwifyFpFn
};
